### PR TITLE
[WIP] Adds new BundleVersion endpoint for file path redirect

### PR DESCRIPTION
**Pull request by eLRuLL**
## Description

LX is presenting a problem of S3 BundleVersions file paths being expired:

    The provided token has expired.

The idea is for blockstore to retrieve file paths to blockstore itself that then redirects to the actual file path

## Test Instructions

Tests included, just run:

    make test

Reviewers:
[ ] @bradenmacdonald 
[ ] @arbrandes
[ ] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @eLRuLL! I've created [OSPR-4987](https://openedx.atlassian.net/browse/OSPR-4987) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://open.edx.org/wp-content/uploads/2019/01/individual-contributor-agreement.pdf)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by natabene** [comment]
@eLRuLL Thank you for the contribution! When you have a chance, could you please send in a signed Contributor Agreement mentioned in the earlier comment? You only need to do this once and we can then review all your future code contributions. Thanks

**Comment by bradenmacdonald** [comment]
@natabene He should be covered under OpenCraft's agreement, as he is part of our team.

**Comment by natabene** [comment]
@bradenmacdonald Got it, thanks for letting me know. I found what the issue was - @eLRuLL  was added as enrull instead of eLRuLL. I will fix it now.

**Comment by eLRuLL** [comment]
ty @natabene. I see the PR is labeled with "waiting on author", please let me know if there is still something I need to do

**Comment by natabene** [comment]
@eLRuLL Usually OpenCraft PRs are first approved by other OC folks, and then by edX. I see that this PR is scheduled to be reviewed by bradenmacdonald and arbrandes. Once they approve, let me know and we will review. Thanks.

File: blockstore/apps/api/v1/views/bundles.py
Comment (arbrandes) [inline blockstore/apps/api/v1/views/bundles.py:36]: Today I learned when to use HTTP 303.  Nice! :+1:

File: blockstore/apps/api/v1/serializers/bundles.py
Comment (arbrandes) [inline blockstore/apps/api/v1/serializers/bundles.py:142]: Do I understand correctly that this will always generate a freshly signed URL, when using the S3 backend?

**Review by arbrandes — CHANGES_REQUESTED** [review]
@eLRuLL, the code looks great!  Aside from an inline question - more a confirmation, really - I do have one request regarding documentation: can you please include a short ADR in the `decisions` directory explaining the motivation behind this change, then an inline comment in `SnapshotField.to_representation()` and/or in the `redirect` view pointing to it?  It may not be readily apparent to future developers why the redirection is being done.

Also, I hate to be a stickler, but can you also please include explicit instructions on how to set up blockstore locally to use S3 (you can refer to the SE-2619 description for that) and then proceed to test that this PR is doing what is intended via curl?  The included unit tests are great, but since this is a rather important change, I'd like to be doubly sure that it won't make anything worse in production.  (Usman being on vacation and all.)

**Comment by bradenmacdonald** [comment]
> include explicit instructions in the PR description on how to set up blockstore locally to use S3

It would be awesome if you could also add that into the blockstore readme. (It's already kind of there in the "production" section, but not as an optional extra step in the "devstack" section).

**Comment by bradenmacdonald** [comment]
@natabene This is related to LabXchange so I'll coordinate review with @ormsbee 

@ormsbee Heads up on this PR, a change we've discussed for a while. Would you like to do the second review+merge here? (If not, I will.)

File: blockstore/apps/api/v1/views/bundles.py
Comment (eLRuLL) [reply blockstore/apps/api/v1/views/bundles.py:37]: me too btw 😄 , @symbolist deserves all the kudos for the recommendation 👍

File: blockstore/apps/api/v1/serializers/bundles.py
Comment (eLRuLL) [reply blockstore/apps/api/v1/serializers/bundles.py:142]: yep, that's the idea

**Comment by eLRuLL** [comment]
@bradenmacdonald @arbrandes the review was really helpful, here some of the improvements:

- Made the redirect endpoint public (I was testing it with LX and, we were not able to connect to it)
- Added more documentation and S3 setup to the README.md
- Changed the local requirements to take constraints as normal dependencies (it wasn't installing necessary packages, like `django-storages`)

For testing with S3 setup:

- Follow S3 steps added to the README.md
- If you have LX already running locally with devstack, check that creating new assets work correctly, bundles are created in the S3 bucket

Please review again, thanks!

**Comment by eLRuLL** [comment]
I was also trying to make this work with "current" assets, but I think I am missing a configuration on edx, as I am getting these errors when trying to re provision:

lms:
```
edx.devstack.lms                  |     version_info = get_bundle_version(bundle_uuid, version_number)
edx.devstack.lms                  |   File "/edx/app/edxapp/edx-platform/openedx/core/lib/blockstore_api/methods.py", line 270, in get_bundle_version
edx.devstack.lms                  |     return api_request('get', version_url)
edx.devstack.lms                  |   File "/edx/app/edxapp/edx-platform/openedx/core/lib/blockstore_api/methods.py", line 49, in api_request
edx.devstack.lms                  |     response.raise_for_status()
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/requests/models.py", line 941, in raise_for_status
edx.devstack.lms                  |     raise HTTPError(http_error_msg, response=self)
edx.devstack.lms                  | requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://edx.devstack.blockstore:18250/api/v1/bundle_versions/a0b9dc6c-1643-4eb6-b223-f61f91fa5667,1
```

blockstore:
```
File "/blockstore/venv/lib/python3.5/site-packages/rest_framework/fields.py", line 454, in get_attribute
    return get_attribute(instance, self.source_attrs)
  File "/blockstore/venv/lib/python3.5/site-packages/rest_framework/fields.py", line 94, in get_attribute
    instance = getattr(instance, attr)
  File "/blockstore/app/blockstore/apps/bundles/models.py", line 217, in staged_draft
    return draft_repo.get(self.uuid)
  File "/blockstore/app/blockstore/apps/bundles/store.py", line 453, in get
    raise StagedDraft.NotFoundError(u"Draft {} not found in {}".format(draft_uuid, self))
blockstore.apps.bundles.store.StagedDraft.NotFoundError: Draft 67c2ae5d-57ec-4159-b813-f3c3c6e74658 not found in <blockstore.apps.bundles.store.DraftRepo object at 0x7fd57d60ecc0>
```

which basically indicates that the snapshot doesn't exist (which is correct, because this is a new S3 bucket), is that something that gets created when we first provision lms/studio? (I was only reprovisioning LX)

@bradenmacdonald @arbrandes any ideas here? As I mentioned, this works for new assets locally, but getting errors for old ones (as they don't exist in S3)

**Comment by eLRuLL** [comment]
I confirmed that a full provision on LX creates the blockstore snapshots btw, so good to go 👍

**Review by arbrandes — APPROVED** [review]
@eLRuLL, apologies for the delay.  I wanted to test this locally with the library-authoring MFE before giving the go-ahead, and I ran into a myriad of issues with my devstack because of the Python 3.8 upgrade.  In any case, :+1:

- [x] I tested this by configuring a local devstack to use S3, then using the library-authoring MFE to create a library with HTML blocks containing assets.
- [x] I read through the code
- N/A I checked for accessibility issues
- [x] Includes documentation

**Comment by bradenmacdonald** [comment]
Why are the tests not running here? We have "build from forks" enabled. I pushed the commit to the `edx` org fork in case it would help, and then it shows up on CircleCI but just says "This project is configured to only run builds that have open pull requests associated with them. Update the 'Only build pull requests' setting to run this build." ?

File: requirements/local.in
Comment (bradenmacdonald) [inline requirements/local.in:1]: Why this change?

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/tests/test_contract.py:449]: Should we remove these `print`s?

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/tests/test_contract.py:298]: Shouldn't we also check that `public_client` can get `response_redirect_url` unmodified, without first parsing it and changing it? I'm assuming it should be an absolute URL.

Which reminds me, the only reason this probably works on devstack or in any environment where blockstore's hostname is different for edxapp vs. public clients is [this hack we have in place](https://github.com/edx/edx-platform/blob/fe550fbd5f7b4ac945d716e41052a79469a7d6a2/openedx/core/lib/blockstore_api/methods.py#L419-L433)

File: blockstore/apps/api/v1/serializers/bundles.py
Comment (bradenmacdonald) [reply blockstore/apps/api/v1/serializers/bundles.py:142]: Isn't that going to make caching performance worse though (always generating a freshly signed URL)?

Before, we'd generate a signed URL then cache it for an hour (or whatever the URLs expiry time was). During that hour, browsers wouldn't try to re-load the asset, they'd just serve it from cache. This should have been working well except for some cache invalidation bug somewhere.

Now, if the URL is changing every time, browsers will never serve the asset from cache. Right?

File: blockstore/apps/bundles/store.py
Comment (bradenmacdonald) [inline blockstore/apps/bundles/store.py:138]: Why did you move the serialization code into this `Snapshot` structure class? Seems inconsistent; I'd rather you just move it to a SnapshotSerializer in the serializers file and update SnapshotField to point to that.

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/tests/test_contract.py:292]: URLs should be hard-coded into tests, not reversed, to ensure the tests will break if we change URLs (breaking API backwards compatibility)

File: blockstore/apps/api/v1/views/bundles.py
Comment (bradenmacdonald) [inline blockstore/apps/api/v1/views/bundles.py:123]: (Ignore this comment) Why a querystring instead of part of the URL? If it's filename-like, putting it in the URL would look cooler.

**Review by bradenmacdonald — COMMENTED** [review]
Sorry if this question has already been asked, but why are we using signed URLs in the first place? If this new view has no authentication, it's not really any different than just handing out unsigned URLs with no expiry time. I guess the only different behavior might be if the bundleversion is deleted, but I don't think we worry about that too much.

Also, does this new URL endpoint make it possible for an anonymous user to discover and fetch different files in the bundle? If they can get one redirect URL like https://blockstore.example.com/bundleversion/uuid/redirect?/xblock/asset/image.png and then construct a new URL like https://blockstore.example.com/bundleversion/uuid/redirect?/xblock/problem-source-code.olx and get the OLX source to a problem (which they weren't supposed to have), then this is a security risk. (In this example, getting the URL to an image used in a problem allows the user to view the problem OLX source and see what the right answer is, cheating on a test etc.)

In other words, knowing the URL to download one file in the bundle shouldn't let you download other files in the bundle by name, unless you are an authorized user. You should have to be given the URL for each file by an authorized user.

**Comment by eLRuLL** [comment]
> Why are the tests not running here? We have "build from forks" enabled. I pushed the commit to the `edx` org fork in case it would help, and then it shows up on CircleCI but just says "This project is configured to only run builds that have open pull requests associated with them. Update the 'Only build pull requests' setting to run this build." ?

No idea really, I thought it was like that, but I just noticed the other PRs do get the CI checks, maybe only for enabled users? maybe can you try enabling that and I can push another commit?

File: requirements/local.in
Comment (eLRuLL) [reply requirements/local.in:1]: > Changed the local requirements to take constraints as normal dependencies (it wasn't installing necessary packages, like django-storages)

IMO the `local.txt` was not built correctly (it was missing requirements)

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (eLRuLL) [reply blockstore/apps/api/v1/tests/test_contract.py:449]: oops, thank you for noticing that

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (eLRuLL) [reply blockstore/apps/api/v1/tests/test_contract.py:298]: it does work with the redirect url, fixing this

File: blockstore/apps/bundles/store.py
Comment (eLRuLL) [reply blockstore/apps/bundles/store.py]: yep, I separated the log to another serializer

File: blockstore/apps/api/v1/tests/test_contract.py
Comment (eLRuLL) [reply blockstore/apps/api/v1/tests/test_contract.py]: makes sense, improved

**Comment by eLRuLL** [comment]
> Sorry if this question has already been asked, but why are we using signed URLs in the first place?

I would really appreciate @ormsbee input here

> If this new view has no authentication, it's not really any different than just handing out unsigned URLs with no expiry time. I guess the only different behavior might be if the bundleversion is deleted, but I don't think we worry about that too much.

I had the same concern, but how should we handle redirect URLs with authentication? append the auth token to the direct url returned would be even worse, or maybe we could come up with expiring tokens, but that will create the same problems as S3 expired token URLs 😅.

I think the secured way would be to enable authentication to this endpoint, but we'll have to change how this is called (from OpenedX) to send authentication when calling those links then.

To make this backward compatible, I think we could add a new parameter to the `retrieve` endpoint so the client decides if it wants to receive a redirect URL for the bundle versions (to which they'd later have to add authentication to retrieve the file link) or if the parameter not passed, they received the link to the file as normal, what do you think @bradenmacdonald ?

**Comment by bradenmacdonald** [comment]
> I had the same concern, but how should we handle redirect URLs with authentication?

The current model is that the download URL for each file must be explicitly given by Blockstore/edxapp, and all URLs eventually expire, so access is completely gated by edxapp.

A more permissive model worth considering is that the download URL for each file must be explicitly given by Blockstore/edxapp, but URLs last forever. edxapp would only be giving out URLs to XBlock asset files anyways, like video files or images used in an XBlock (OLX file URLs would never be given out), so I think this is both acceptable and performant.

This PR's proposed change would allow users to construct a download URL for any file in a bundle (once they had any other file), and the URLs would last forever. This would allow things like finding and reading the OLX source of a problem to determine the correct answer, so I think it is not acceptable.

Let's discuss which of the first two options we should go with, or if there are other options or considerations I haven't mentioned here.

**Comment by natabene** [comment]
@eLRuLL @bradenmacdonald Have you had a chance to discuss how to move this forward?

**Comment by natabene** [comment]
@bradenmacdonald Would you like to review this as a Core Commmitter?

**Comment by eLRuLL** [comment]
I think this could be closed or maybe use already the conversation to hear from @ormsbee on what would be his recommendations on how to proceed

**Comment by bradenmacdonald** [comment]
@natabene @eLRuLL I suggest closing this for now. We did some discovery and there are potential content security concerns with this approach. We can always revisit this once Blockstore's code is moved into the LMS, but for now I don't think this is the right path forward.

**Comment by eLRuLL** [comment]
yep, I definitely agree, closing this

**Comment by edx-webhook** [comment]
@eLRuLL Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/blockstore&entry.867055334=https://github.com/edx/blockstore/pull/87&entry.1484655318=https://github.com/eLRuLL&entry.752974735=2020-09-19+15:39&entry.1917517419=2021-01-14+01:22&entry.2133058324=No) so we can improve your experience in the future.

**Pull request closed without merging**
